src/main/java/com/iota/iri/utils/thread/BoundedScheduledExecutorService.java

```diff
@@ -0,0 +1,931 @@
+
+        return wrappedFuture.delegate(futureFactory.create(wrappedCallable));
+    }
+
+    /**
+     * This is a utility method that wraps the task and the resulting {@link ScheduledFuture} in a single call.<br />
+     * <br />
+     * It creates the {@link WrappedScheduledFuture} and the wrapped task and starts the execution of the task by
+     * delegating the launch through the {@link FutureFactory}.<br />
+     *
+     * @param futureFactory the lambda that returns the original "unwrapped" future from the wrapped task
+     * @param task the task that shall be wrapped to clean up its reserved resources upon completion
+     * @param taskDetails metadata holding the relevant information of the task
+     * @param <V> the return type of the task
+     * @return the {@link WrappedFuture} that allows to interact with the task
+     */
+    private <V> ScheduledFuture<V> wrapScheduledFuture(FutureFactory<ScheduledFuture<V>, Runnable> futureFactory,
+            Runnable task, TaskDetails taskDetails) {
+
+        WrappedScheduledFuture<V> wrappedFuture = new WrappedScheduledFuture<>(taskDetails);
+        Runnable wrappedRunnable = wrapTask(task, taskDetails, wrappedFuture);
+
+        return wrappedFuture.delegate(futureFactory.create(wrappedRunnable));
+    }
+
+    /**
+     * This is a utility method that wraps a {@link Collection} of {@link Callable}s to make them free their resources
+     * upon completion.<br />
+     * <br />
+     * It simply iterates over the tasks and wraps them one by one by calling
+     * {@link #wrapTask(Callable, TaskDetails)}.<br />
+     *
+     * @param tasks list of jobs that shall be wrapped
+     * @param <T> the type of the values returned by the {@link Callable}s
+     * @return wrapped list of jobs that will free their reserved resources upon completion
+     */
+    private <T> Collection<? extends Callable<T>> wrapTasks(Collection<? extends Callable<T>> tasks,
+            TaskDetails taskDetails) {
+
+        List<Callable<T>> wrappedTasks = new ArrayList<>(tasks.size());
+        for (Callable<T> task : tasks) {
+            wrappedTasks.add(wrapTask(task, taskDetails));
+        }
+
+        return wrappedTasks;
+    }
+
+    /**
+     * This method checks if we have enough resources to schedule the given amount of tasks and reserves the space if
+     * the check is successful.<br />
+     * <br />
+     * The reserved resources will be freed again once the tasks finish their execution or when they are cancelled
+     * through their corresponding {@link Future}.<br />
+     *
+     * @param requestedJobCount the amount of tasks that shall be scheduled
+     * @return true if we could reserve the given space and false otherwise
+     */
+    private boolean reserveCapacity(int requestedJobCount) {
+        if (tasksQueued.addAndGet(requestedJobCount) <= capacity) {
+            return true;
+        } else {
+            tasksQueued.addAndGet(-requestedJobCount);
```

Comment:
I don't get why this is here..
I didn't notice you are freeing resources due to this...

Reply:
ohh of course,
very obvious the second time I look at it :-)